core/src/main/java/com/yahoo/oak/Block.java

```diff
@@ -37,9 +37,19 @@ void setID(int id) {
     // Block manages its linear allocation. Thread safe.
     // The returned buffer doesn't have all zero bytes.
     boolean allocate(Slice s, int size) {
-        int now = allocated.getAndAdd(size);
-        if (now + size > this.capacity) {
-            allocated.getAndAdd(-size);
+        assert size >= 0; // avoid negative size
+        int now = 0;
+        boolean invalid = false;
+        synchronized (allocated) { // this is suboptimal, will optimize
+            // long to manage integer overflow or use AtomicLong for single getAndAdd for valid scenario
+            long next = 0L + allocated.get() + size;
+            if (next > this.capacity) {
+                invalid = true; // use flag instead of exception to reduce time spent inside synchronized block
+            } else {
+                now = allocated.getAndAdd(size);
+            }
+        }
+        if (invalid) {
             throw new OakOutOfMemoryException(String.format("Block %d is out of memory", id));
         }
         s.associateBlockAllocation(id, now, size, buffer);
```

Comment:
lets consider the following scenario:

1. capacity = 10
2. allocated = 0
3. 4 threads (R1, R2, R3, R4)
4. block id = B1
5. Different execution times of 4 threads are tracked as T1, T2, T3, T4
6. Assume there is only one core available for execution

## Time sequenced operations:
### T1.1: R1.allocate(size = 11)
1. now = 0
2. allocated = 11 (0 + 11)
3. if (0 + 11 > 10) - true (invalid)
4. context switch

### T2.1: R2.allocate(size = 1)
1. now = 11
2. allocated = 12 (11 + 1)
3. if (11 + 1 > 10) - true (invalid, though this is an ambiguous state. still its fine to proceed with the analysis)
4. context switch

### T1.2 : R1.allocate(size = 11): resumes from T1.1
1. now = 0
2. allocated = 12
3. allocated.add(-11)
4. allocated = 1
5. exception due to true in if check in T1.1
6. completed

### T3.1 : R3.allocate(size = 2)
1. now = 1
2. allocated = 3 (1 + 2)
3. if (3 > 10) - false (valid)
4. s.associateBlockAllocation(B1, offset = 1 , size = 2 , buffer) - (uses: 1 and 2)
5. completed

### T2.2: R2.allocate(size = 1): resumes from T2.1
1. now = 11
2. allocated = 3
3. allocated.add(-1) (due to T2.1 check of 11 + 1 > 10)
4. allocated = 2
5. exception due to true in if check in T2.1
6. completed
7. now offset = 0 is in an ambiguous state

### T4.1: R4.allocate(size = 1)
1. now = 2 (from allocated)
2. allocated = 3 (2 + 1)
3. if (3 > 10) - false (valid)
4. s.associateBlockAllocation(B1, offset = 2 , size = 1 , buffer) - (uses: 2)
completed
5. now offset = 2 is clearly in violated state with both R3.allocate(size = 2) and R4.allocate(size = 1) sharing/overlapping at offset 2

can you clarify whether this is a valid scenario?